keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,249 @@
+  to the **kube-migrator controller** `deployment`. Thus, the old `migration`s
+  are deleted with the old `deployment` in the first step.
+* uses the same kubeconfig file as **kubectl** does.
+
+The control loop of **kube-migrator controller** does the following:
+* watches the instances of the `migration` CR, adds them to a [workqueue][].
+* syncs one `migration` at a time to avoid overloading the apiserver,
+  * if `migration.status.completed == false`, it creates a **migration worker**
+    `job` to migrate the resource type. The `ownerReference` of the `job` is set
+    to the `migration`.
+  * waits until the **migration worker** `job` finishes, updates
+    `migration.status.completed` to `true`, and deletes the `job`.
+
+The **migrator worker** `job` runs the equivalence of `oc adm migrate storage
+--include=<resource type>` to migrate a resource type. The **migrator worker**
+uses API chunking to retrieve partial lists of a resource type and thus can
+migrate a small chunk at a time. It stores the [continue token] in the owner
+`migration.status.continueToken`. With the inconsistent continue token
+introduced in [#67284][], the **migrator worker** does not need to worry about
+expired continue token.
+
+[list]:https://github.com/openshift/origin/blob/2a8633598ef0dcfa4589d1e9e944447373ac00d7/pkg/oc/cli/admin/migrate/storage/storage.go#L120-L184
+[workqueue]:https://github.com/kubernetes/client-go/tree/master/util/workqueue
+[#67284]:https://github.com/kubernetes/kubernetes/pull/67284
+[API section]:#alpha-api
+
+The cluster admin can run the `kubectl get migrations -o
+jsonpath="{.status.completed}"` to see if all migrations have completed.
+
+We will support a `kube-migrator migrate --resources=<list of resources>`
+command. It (re)creates `migration` CRs for the resources to trigger their
+migration. This command facilitates manually migrating custom resources and
+aggregated resources.
+
+### Alpha API
+
+We introduce the `migration` API to record the intention and the progress of a
+migration. The API will be a CRD defined in the `migration.k8s.io` group.
+
+Read the [workflow section][] to understand how the API is used.
+
+```golang
+type StorageVersionMigration struct {
+  metav1.TypeMeta
+  // metadata.name should be "<resource>.<group>" of the resource being
+  // migrated.
+  // Because resouces can be served at different groups, in the future version,
+  // metadata.name can be hash(etcd instance, etcd path prefix).
+  metav1.ObjectMeta
+  Spec StorageVersionMigrationSpec
+  Status StorageVersionMigrationStatus
+}
+
+type StorageVersionMigrationSpec {
+  // Resource is the resource that is being migrated. The migrator sends
+  // requests to the endpoint serving the Resource.
+  // Immutable.
+  Resource GroupVersionResource
+}
+
+type StorageVersionMigrationStatus {
+  // Completed means the migration of this resource has completed.
+  Completed bool
```

Comment:
I expected status to include conditions which are fairly standard in the API and have built in scripting functionality.

Reply:
Do you mean making it a enum supporting states like Running, Succeeded, Failed? I agree it should be an enum instead of an boolean.

I'm not sure what you mean by "built in scripting functionality"?

Reply:
I'll start with adding `.status.phase`.

Reply:
Ok, I've converted to conditions.